File: 2016/js/day2.js
window.AdventOfCode.Day2 = function( input )
{
	input = input.split( '\n' );
	
	// Part 1
	let keypad =
	[
		[ 1, 2, 3 ],
		[ 4, 5, 6 ],
		[ 7, 8, 9 ],
	];
	
	let Clamp = function( i )
	{
		return i < 0 ? 0 : i > 2 ? 2 : i;
	}
	
	let position = { x: 1, y: 1 };
	let part1 = "";
	
	for( let line of input )
	{
		for( let i of line )
		{
			switch( i )
			{
				case 'L': position.x--; break;
				case 'R': position.x++; break;
				case 'U': position.y--; break;
				case 'D': position.y++; break;
			}
			
			position.x = Clamp( position.x );
			position.y = Clamp( position.y );
		}
		
		part1 += keypad[ position.y ][ position.x ];
	}
	
	// Part 2
	keypad =
	[
		[ null, null,   1, null,  null ],
		[ null,    2,   3,    4,  null ],
		[    5,    6,   7,    8,     9 ],
		[ null,  'A',  'B',  'C', null ],
		[ null, null,  'D', null, null ],
	];
	
	let IsValidPosition = function()
	{
		if( !keypad[ position.y ] )
		{
			return false;
		}
		
		return !!keypad[ position.y ][ position.x ];
	};
	
	position = { x: 0, y: 2 };
	let part2 = "";
	
	for( let line of input )
	{
		for( let i of line )
		{
			const currentPosition = { x: position.x, y: position.y };
			
			switch( i )
			{
				case 'L': position.x--; break;
				case 'R': position.x++; break;
				case 'U': position.y--; break;
				case 'D': position.y++; break;
			}
			
			if( !IsValidPosition() )
			{
				position = currentPosition;
			}
		}
		
		part2 += keypad[ position.y ][ position.x ];
	}
	
	return [ part1, part2 ];
};

File: 2025/Answers/Solutions/Day2.cs
using System;

namespace AdventOfCode;

[Answer(2)]
public class Day2 : IAnswer
{
	public Solution Solve(string inputStr)
	{
		var part1 = 0L;
		var part2 = 0L;
		var input = inputStr.AsSpan();
		Span<char> s = stackalloc char[19];

		foreach (var numberRange in input.Split(','))
		{
			var range = input[numberRange];
			var dashIndex = range.IndexOf('-');
			var start = range[..dashIndex].ParseLong();
			var end = range[(dashIndex + 1)..].ParseLong();

			for (var n = start; n <= end; n++)
			{
				var t = Check2(n, s);

				if (t > 0)
				{
					if (t % 2 == 0)
					{
						part1 += n;
					}

					part2 += n;
				}
			}
		}

		return new(part1.ToString(), part2.ToString());
	}

	private int Check2(long n, Span<char> s)
	{
		n.TryFormat(s, out var length);

		for (var check = 1; check <= length / 2; check++)
		{
			if (length % check != 0)
			{
				continue;
			}

			var first = s[..check];
			var repetitions = length / check;
			var allMatch = true;

			for (var i = check; i < length; i += check)
			{
				if (!first.SequenceEqual(s[i..(i + check)]))
				{
					allMatch = false;
					break;
				}
			}

			if (allMatch)
			{
				return repetitions;
			}
		}

		return 0;
	}
}
